File: 15.AhoCorasickAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio      */
/* -----------------------------------------------------------------------  */

import java.io.*;
import java.util.*;

class AhoCorasick {
    // Node structure for the trie
    static class Node {
        Map<Character, Node> next = new HashMap<>(); // transitions
        Node fail;                                   // failure link
        List<Integer> out = new ArrayList<>();       // store pattern indexes
    }

    private Node root = new Node();

    // Step 1: Insert patterns into trie
    public void buildTrie(String[] patterns) {
        for (int idx = 0; idx < patterns.length; idx++) {
            String word = patterns[idx];
            Node cur = root;
            for (char ch : word.toCharArray()) {
                cur = cur.next.computeIfAbsent(ch, k -> new Node());
            }
            cur.out.add(idx); // store index of pattern ending here
        }
    }

    // Step 2: Build failure links using BFS
    public void buildFailureLinks() {
        Queue<Node> q = new LinkedList<>();
        // Initialize fail links of depth-1 nodes
        for (Node nxt : root.next.values()) {
            nxt.fail = root;
            q.add(nxt);
        }

        while (!q.isEmpty()) {
            Node cur = q.poll();
            for (Map.Entry<Character, Node> entry : cur.next.entrySet()) {
                char ch = entry.getKey();
                Node nxt = entry.getValue();
                q.add(nxt);

                Node f = cur.fail;
                while (f != null && !f.next.containsKey(ch)) {
                    f = f.fail;
                }
                nxt.fail = (f == null) ? root : f.next.get(ch);
                if (nxt.fail != null) {
                    nxt.out.addAll(nxt.fail.out);
                }
            }
        }
    }

    // Step 3: Search text
    public List<String> search(String text, String[] patterns) {
        List<String> res = new ArrayList<>();
        Node cur = root;

        for (int i = 0; i < text.length(); i++) {
            char ch = text.charAt(i);

            while (cur != null && !cur.next.containsKey(ch)) {
                cur = cur.fail; // follow failure links
            }
            if (cur == null) {
                cur = root;
                continue;
            }

            cur = cur.next.get(ch);

            for (int idx : cur.out) {
                String word = patterns[idx];
                int pos = i - word.length() + 1;
                res.add("Pattern '" + word + "' found at position " + pos);
            }
        }
        return res;
    }

    // Driver Code
    public static void main(String[] args) throws IOException {
        String[] patterns = {"he", "she", "his", "hers"};
        String text = "ahishers";

        AhoCorasick ac = new AhoCorasick();
        ac.buildTrie(patterns);
        ac.buildFailureLinks();

        List<String> matches = ac.search(text, patterns);

        System.out.println("Matches found:");
        for (String s : matches) {
            System.out.println(s);
        }
    }
}

File: 15.AhoCorasickAlgo.py
# The Aho-Corasick algorithm is a powerful string searching algorithm
# that efficiently finds all occurrences of multiple patterns in a text.
# It builds a finite state machine (trie + failure links) to match all patterns simultaneously in linear time.

# Explanation

# First, we insert all patterns into a trie (prefix tree).

# Then, we build failure links that help us skip to the longest suffix which is
# also a prefix when mismatches occur.

# Finally, we process the input text character by character, following trie edges, a
# nd use failure links to report all matches.

# This allows us to search multiple patterns at once in O(N + M + Z) time, where:

# N = length of text

# M = total length of all patterns

# Z = total number of matches

from collections import deque

class AhoCorasick:
    def __init__(self, patterns):
        self.trie = {}
        self.output = {}
        self.fail = {}
        self.build_trie(patterns)
        self.build_failure_links()

    def build_trie(self, patterns):
        new_state = 0
        for idx, word in enumerate(patterns):
            state = 0
            for ch in word:
                if (state, ch) not in self.trie:
                    new_state += 1
                    self.trie[(state, ch)] = new_state
                state = self.trie[(state, ch)]
            self.output.setdefault(state, []).append(idx)

    def build_failure_links(self):
        q = deque()
        self.fail[0] = 0
        for (state, ch), nxt in list(self.trie.items()):
            if state == 0:
                self.fail[nxt] = 0
                q.append(nxt)

        while q:
            r = q.popleft()
            for (state, ch), nxt in list(self.trie.items()):
                if state == r:
                    q.append(nxt)
                    f = self.fail[state]
                    while (f, ch) not in self.trie and f != 0:
                        f = self.fail[f]
                    self.fail[nxt] = self.trie.get((f, ch), 0)
                    self.output.setdefault(nxt, [])
                    self.output[nxt] += self.output.get(self.fail[nxt], [])

    def search(self, text, patterns):
        state = 0
        results = []
        for i, ch in enumerate(text):
            while (state, ch) not in self.trie and state != 0:
                state = self.fail[state]

            state = self.trie.get((state, ch), 0)

            for pattern_idx in self.output.get(state, []):
                word = patterns[pattern_idx]
                pos = i - len(word) + 1
                results.append((word, pos))
        return results


patterns = ["he", "she", "his", "hers"]
text = "ahishers"

ac = AhoCorasick(patterns)
matches = ac.search(text, patterns)

print("Matches found:")
for word, pos in matches:
    print(f"Pattern `{word}` found at position {pos}")
